### PR TITLE
Multi-Tenant support in the REST Proxy

### PR DESCRIPTION
**Pull request by HowardWolosky**
Some companies use multiple developer accounts for different reasons,
and want to be able to use a single Proxy server to handle requests
for all of these different accounts.

To enable this scenario, ProxyManager has been refactored.

* `EndpointInfo` used to be a shallow object which stored a few
  properties that ProxyManager used to perform the actual proxy
  request redirection.  It is now called `Endpoint` and is a
  self-contained object that, once initialized with the necessary
  configuration, can perform an authenticated proxy request.

* `ProxyManager` has become thinner, and is now simply reponsible
  for choosing the correct `Endpoint` based on the incoming request,
  and then letting the selected `Endpoint` handle the acutal
  request.

    * It stores the Endpoints in two different dictionaries: one
      indexed by TenantId, the other by TenantFriendlyName, allowing
      for quicker lookup in either usage scenario.

* Most of the service configuration properties around Tenants have
  been removed, replaced with just two properties: `DefaultTenantId`
  and `EndpointJsonConfig`.  `DefaultTenantId` indicates the
  `TenantId` that a request should use if the proxy request didn't
  explicitly speify one.  `EndpointJsonConfig` is a JSON array of
  `Endpoint` objects, enabling a deployed Proxy to handle any number
  of tenants and endpoints.

* At the time the WebApi is being registered, the ProxyManager will
  be statically configured with the endpoints / default TenantId as
  specified in the service configuration.

* When `Set-StoreBrokerAuthentication -UseProxy` is called, users can
  now optionally specify `-TenantId` (with the full TenantId of the
  tenant that they want the proxy to use) or `-TenantName` (a friendly
  name for that TenantId).  These values will be passed on to the Proxy
  using additional headers, similar to how `UseINT` is being handled
  today.

Additionally:
 * The RootController has been expanded to support 2 additional
   levels of REST API commands to better handle future API improvements
   which may leverage that deeper nesting of commands within the URI.

### CODE REVIEW

File: Documentation/RESTPROXY.md [hunk 1]
```diff
@@ -335,6 +336,56 @@ First, get the IP address of the machine you want to remote into.
...
 service account username and password.
 
+### Endpoint JSON Configuration
+
+The Proxy supports managing multiple tenants in a single Proxy instance.  Because of this,
+the tenant-related configuration for the proxy is done within a JSON object that is stored as
+a string property (`EndpointJsonConfig`) within the AzureService configuration
+(`ServiceConfiguration.Cloud.cscfg`).
+
+This JSON consists of an **array** of `Endpoint`s.
```
Comment (danbelcher-MSFT) [inline Documentation/RESTPROXY.md]: `Endpoint` objects?  Looks odd when rendered

Comment (HowardWolosky) [inline Documentation/RESTPROXY.md]: sure. changed.

File: Documentation/RESTPROXY.md [hunk 2]
```diff
@@ -457,12 +508,14 @@ one that you just created.
...
-  * DomainJoinAccountCertThumbprint
-  * ClientSecretProdCertThumbprint
-  * ClientSecretIntCertThumbprint
-  * And you'll need to update the `certificate` entry for the one named
-  `StoreBroker Proxy Authentication`
+  * `LocalAdminAccountCertThumbprint`
+  * `DomainJoinAccountCertThumbprint`
+
+You should also update the `clientSecretCertificateThumbprint` for the PROD and INT endpoints
+within your [`EndpointJsonConfig`](#tenant-json-configuration).
```
Comment (danbelcher-MSFT) [inline Documentation/RESTPROXY.md]: #endpoint-json-configuration?

File: Documentation/RESTPROXY.md [hunk 3]
```diff
@@ -457,12 +508,14 @@ one that you just created.
...
+
+You'll also need to update the `certificate` entry for the one named
```
Comment (danbelcher-MSFT) [inline Documentation/RESTPROXY.md]: nit: I would remove "the one named"

File: Documentation/RESTPROXY.md [hunk 4]
```diff
@@ -644,9 +697,25 @@ There are a couple things to note when doing local development:
     `ProxyManager.GetClientSecret`.   
 
  * When developing locally, you will need to tell StoreBroker to use a different endpoint for
-   the proxy.  You can achieve that by running
-
-       Set-StoreBrokerAuthentication -UseProxy -ProxyEndpoint <local endpoint>
+   the proxy.  By default, this local endpoint value will be `http://localhost:1034`.  You can
+   achieve that by running:
+
+       Set-StoreBrokerAuthentication -UseProxy -ProxyEndpoint 'http://localhost:1034'
```
Comment (danbelcher-MSFT) [inline Documentation/RESTPROXY.md]: This isn't rendering as a code-block

File: Documentation/RESTPROXY.md [hunk 5]
```diff
@@ -691,6 +760,10 @@ specific differences:
    * If you want to use _INT_ instead of _PROD_, provide this additional header value:
```
Comment (danbelcher-MSFT) [inline Documentation/RESTPROXY.md]: Should we change this comment, given that only the first-party account can use INT?

File: Documentation/RESTPROXY.md [hunk 6]
```diff
@@ -691,6 +760,10 @@ specific differences:
...
      `UseINT = true`
 
+   * If the proxy being used is multi-tenant aware, you can either pass in the tenantId in the
```
Comment (danbelcher-MSFT) [inline Documentation/RESTPROXY.md]: nit: tenantId should either be `tenantId` or re-written as natural English, ie 'tenant ID'

File: Documentation/RESTPROXY.md (hunk 2 above)
Comment (HowardWolosky) [inline Documentation/RESTPROXY.md]: fixed

File: Documentation/RESTPROXY.md (hunk 3 above)
Comment (HowardWolosky) [inline Documentation/RESTPROXY.md]: done

File: Documentation/RESTPROXY.md (hunk 4 above)
Comment (HowardWolosky) [inline Documentation/RESTPROXY.md]: trying again

File: RESTProxy/Models/Endpoint.cs [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+        }
+
+        /// <summary>
+        /// Describes the type of endpoint that the request will be proxy-ed through.
+        /// </summary>
+        public enum EndpointType
+        {
+            /// <summary>
+            /// The production (live) endpoint.
+            /// Changes made via this endpoint will be affect customers.
```
Comment (danbelcher-MSFT) [inline RESTProxy/Models/Endpoint.cs]: nit: remove 'be'

File: RESTProxy/Models/Endpoint.cs [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+        /// <returns>
+        /// The AccessToken that can be used for API requests on this Endpoint.
+        /// </returns>
+        private async Task<string> GetAccessTokenAsync()
+        {
+            if (this.ValidThru.CompareTo(DateTime.Now) <= 0)
+            {
+                using (await this.AsyncLock.LockAsync())
+                {
+                    // To otimize the normal case, we won't acquire a lock when initially checking ValidThru,
```
Comment (danbelcher-MSFT) [inline RESTProxy/Models/Endpoint.cs]: typo

File: RESTProxy/Models/Endpoint.cs [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+                errorResponse = new HttpResponseMessage(HttpStatusCode.Unauthorized)
+                {
+                    Content = new StringContent(response, Encoding.UTF8, ProxyManager.JsonMediaType)
+                };
+
+                return false;
+            }
+        }
+    }
+}
```
Comment (danbelcher-MSFT) [inline RESTProxy/Models/Endpoint.cs]: newline at end of file

File: Documentation/RESTPROXY.md (hunk 5 above)
Comment (HowardWolosky) [inline Documentation/RESTPROXY.md]: Added a note within there that explains that better.

File: Documentation/RESTPROXY.md (hunk 6 above)
Comment (HowardWolosky) [inline Documentation/RESTPROXY.md]: done

File: RESTProxy/Models/Endpoint.cs (hunk 7 above)
Comment (HowardWolosky) [inline RESTProxy/Models/Endpoint.cs]: fixed

File: RESTProxy/Models/Endpoint.cs (hunk 8 above)
Comment (HowardWolosky) [inline RESTProxy/Models/Endpoint.cs]: fixed

File: RESTProxy/Models/Endpoint.cs (hunk 9 above)
Comment (HowardWolosky) [inline RESTProxy/Models/Endpoint.cs]: fixed

**Pull request merged**